pkg/storegateway/bucket_index_reader.go

```diff
@@ -268,6 +224,84 @@ func (r *bucketIndexReader) expandedPostings(ctx context.Context, ms []*labels.M
 	return ps, nil
 }
 
+// toPostingGroups returns a set of labels for which to look up postings lists. It guarantees that
+// each postingGroup's keys exist in the index. The order of the returned labels
+// is the same as iterating through the posting groups and for each group adding all keys.
+func toPostingGroups(ms []*labels.Matcher, indexhdr indexheader.Reader) ([]postingGroup, []labels.Label, error) {
+	var (
+		rawPostingGroups = make([]rawPostingGroup, 0, len(ms))
+		allRequested     = false
+		hasAdds          = false
+	)
+
+	// NOTE: Derived from tsdb.PostingsForMatchers.
+	for _, m := range ms {
+		// Each group is separate to tell later what postings are intersecting with what.
+		rawPostingGroups = append(rawPostingGroups, toPostingGroup(m))
+	}
+
+	// First iterate through the non-lazy groups. We can check whether their requested values exist
+	// in the index. After that we can check the lazy groups. The lazy groups require calling LabelValues
+	// which is more expensive, so we leave them for last.
+	sort.Slice(rawPostingGroups, func(i, j int) bool {
+		return (!rawPostingGroups[i].isLazy && rawPostingGroups[j].isLazy) ||
+			(len(rawPostingGroups[i].keys) < len(rawPostingGroups[j].keys))
+	})
```

Comment:
I would suggest:

```suggestion
	sort.Slice(rawPostingGroups, func(i, j int) bool {
		ri, rj := rawPostingGroups[i], rawPostingGroups[j]
		if ri.isLazy != rj.isLazy {
			return !ri.isLazy
		}
		
		if len(ri.keys) != len(rj.keys) {
			return len(ri.keys) < len(rj.keys)
		}
		
		return ri.labelName < rj.labelName
	})
```

Because:
- This sorts lazy in both ways (in your example, if j is lazy but i is not, you still fall back into comparing the keys length).
- Then we compare the keys length.
- Then we sort by label name, which makes it more deterministic (not entirely though, since we might have multiple matchers for same label) and puts the `__name__` label matcher the first one (which could be useful under _some scenarios_ since it would just discard that a given metric isn't in this block).